### PR TITLE
Make Encoding.Unicode.GetString faster on 64-bit platforms

### PR DESCRIPTION
**Pull request by jkotas**
UnicodeEncoding was taking fast path only when both source and destination were 8-byte aligned. It was not the case for Unicode.GetString.

The fix is to just check the source alignment as fast path precondition, and use unaligned write for destination.

Also removed alternative fast path that is not needed anymore, tweaked local variable assignments for better register allocation, and fixed performance issue introduced by recent big endian portability fix.

Fixes https://github.com/dotnet/corefx/issues/24192

### CODE REVIEW

**Comment by jkotas** [comment]
Improvements (before/after):

System.Text.Tests.Perf_Encoding.GetString(size:   10000, encName: "unicode") | 12.7
-- | --
System.Text.Tests.Perf_Encoding.GetString(size:   1000000, encName: "unicode") | 7.3
System.Text.Tests.Perf_Encoding.GetString(size:   128, encName: "unicode") | 4.9
System.Text.Tests.Perf_Encoding.GetString(size:   16, encName: "unicode") | 1.8
System.Text.Tests.Perf_Encoding.GetString(size:   256, encName: "unicode") | 7.2
System.Text.Tests.Perf_Encoding.GetString(size:   32, encName: "unicode") | 2.7
System.Text.Tests.Perf_Encoding.GetString(size:   512, encName: "unicode") | 8.3
System.Text.Tests.Perf_Encoding.GetString(size:   64, encName: "unicode") | 4.0
  |  
System.Text.Tests.Perf_Encoding.GetByteCount(size:   10000, encName: "unicode") | 2.4
System.Text.Tests.Perf_Encoding.GetByteCount(size:   1000000, encName: "unicode") | 2.4
System.Text.Tests.Perf_Encoding.GetByteCount(size:   128, encName: "unicode") | 1.3

File: src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs [hunk 1]
```diff
@@ -6,11 +6,17 @@
 // Don't override IsAlwaysNormalized because it is just a Unicode Transformation and could be confused.
 //
 
+// This define can be used to turn off the fast loops. Useful for finding whether
+// the problem is fastloop-specific.
+#define FASTLOOP
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs]: To match UTF8Encoding and avoid double negation.

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 corefx_baseline

File: src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs [hunk 2]
```diff
@@ -695,14 +699,14 @@ internal override unsafe int GetByteCount(char* chars, int count, EncoderNLS enc
                 if (ch == 0)
                 {
                     // No fallback, maybe we can do it fast
-#if !NO_FAST_UNICODE_LOOP
+#if FASTLOOP
                     // If endianess is backwards then each pair of bytes would be backwards.
                     if ( (bigEndian ^ BitConverter.IsLittleEndian) && 
-#if BIT64           // 64 bit CPU needs to be long aligned for this to work, 32 bit CPU needs to be 32 bit aligned
-                        (unchecked((long)chars) & 7) == 0 && (unchecked((long)bytes) & 7) == 0 &&
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs]: >(unchecked((long)bytes) & 7) == 0  [](start = 61, length = 34)

we don't care about the bytes alignment here?

File: src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs [hunk 3]
```diff
@@ -452,7 +455,8 @@ internal override unsafe int GetByteCount(char* chars, int count, EncoderNLS enc
 
                                     // If they happen to be high/low/high/low, we may as well continue.  Check the next
                                     // bit to see if its set (low) or not (high) in the right pattern
-                                    if (((0xfc00fc00fc00fc00 & *longChars) ^ highLowPatternMask) != 0)
+                                    if ((0xfc00fc00fc00fc00 & *longChars) !=
+                                            (BitConverter.IsLittleEndian ? (ulong)0xdc00d800dc00d800 : (ulong)0xd800dc00d800dc00))
```
Comment (tarekgh) [inline src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs]: >(BitConverter.IsLittleEndian ? (ulong)0xdc00d800dc00d800 : (ulong)0xd800dc00d800dc00) [](start = 44, length = 85)

can we just cache this value and avoid the checks when it gets used?

**Comment by tarekgh** [comment]
Modulo my questions, LGTM

**Comment by tarekgh** [comment]
@dotnet-bot test Windows_NT x64 Checked corefx_baseline
@dotnet-bot test Windows_NT x86 Checked corefx_baseline

File: src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs (hunk 3 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs]: The IsLittleEndian check is optimized out by the JIT.

The current code that is caching it in readonly variable has worse performance. This caching was introduced just recently in #18254. This change is fixing the performance regression introduced by #18254.

File: src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs (hunk 2 above)
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Text/UnicodeEncoding.cs]: The case of mis-aligned `bytes` is taken care of by using `Unsafe.WriteUnaligned` for writing to them.

**Comment by jkotas** [comment]
System.Transactions.Tests.AsyncTransactionScopeTests.VerifyNestedTxScope failures are https://github.com/dotnet/corefx/issues/29940

**Pull request merged**